subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -414,27 +402,31 @@ private static Type extractServiceType(Type mayBeFactoryType) {
          * Collects all services of the given type.
          */
         void getAll(LookupContext context, Class<?> serviceType, List<ServiceProvider> result);
-
-        boolean hasService(Class<?> type);
     }
 
     private class OwnServices implements Provider {
-        private List<Provider> providers;
-        private Set<Class<?>> serviceTypes;
+        private final Map<Class<?>, Collection<Provider>> providersByType = new HashMap<Class<?>, Collection<Provider>>();
+        private final CompositeStoppable stoppable = CompositeStoppable.stoppable();
 
         @Override
         public ServiceProvider getFactory(LookupContext context, Class<?> type) {
-            if (!hasService(Factory.class)) {
+            Collection<Provider> providers = getProvidersByType(Factory.class);
+            if (providers.isEmpty()) {
                 return null;
             }
-            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>();
+            if (providers.size() == 1) {
+                return providers.iterator().next().getFactory(context, type);
+            }
+
+            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>(providers.size());
```

Comment:
:o: Would there be an interest in not pessimistically creating a list here? Make it null, assume a single candidate. If the candidate is `null`, return `null`. If there's already one, create the list.

Reply:
This never actually shows up in the profiler results, so I didn't go for it. In 99% of the cases we look for non-parametrized types, which are covered by the if-statements above.